src/rapids_singlecell/preprocessing/_qc.py

```diff
@@ -0,0 +1,434 @@
+            gene_ex = cp.zeros((X_part.shape[1]), dtype=cp.int32)
+            if not X_part.flags.c_contiguous:
+                X_part = cp.asarray(X_part, order="C")
+            block = (16, 16)
+            grid = (
+                int(math.ceil(X_part.shape[0] / block[0])),
+                int(math.ceil(X_part.shape[1] / block[1])),
+            )
+            sparse_qc_dense_genes(
+                grid,
+                block,
+                (
+                    X_part,
+                    sums_genes,
+                    gene_ex,
+                    X_part.shape[0],
+                    X_part.shape[1],
+                ),
+            )
+            return cp.vstack([sums_genes, gene_ex.astype(X_part.dtype)])
+    else:
+        raise ValueError(
+            "Please use a cupy csr_matrix or cp.ndarray. csc_matrix are not supported with dask."
+        )
+
+    blocks = X.to_delayed().ravel()
+    cell_blocks = [
+        da.from_delayed(
+            __qc_calc_1(block),
+            shape=(2, X.chunks[0][ind]),
+            dtype=X.dtype,
+            meta=cp.array([]),
+        )
+        for ind, block in enumerate(blocks)
+    ]
+
+    blocks = X.to_delayed().ravel()
+    gene_blocks = [
+        da.from_delayed(
+            __qc_calc_2(block),
+            shape=(2, X.shape[1]),
+            dtype=X.dtype,
+            meta=cp.array([]),
+        )
+        for ind, block in enumerate(blocks)
+    ]
+    sums_cells, cell_ex = da.hstack(cell_blocks)
+    sums_genes, gene_ex = da.stack(gene_blocks, axis=1).sum(axis=1)
+
+    sums_cells, cell_ex, sums_genes, gene_ex = dask.compute(
+        sums_cells, cell_ex, sums_genes, gene_ex
+    )
+
+    return (
+        sums_cells.ravel(),
+        sums_genes.ravel(),
+        cell_ex.ravel().astype(cp.int32),
+        gene_ex.ravel().astype(cp.int32),
+    )
+
+
+def _second_pass_qc(X, mask):
+    if isinstance(X, DaskArray):
+        return _second_pass_qc_dask(X, mask)
+    sums_cells_sub = cp.zeros(X.shape[0], dtype=X.dtype)
+    if sparse.issparse(X):
+        if sparse.isspmatrix_csr(X):
+            from ._kernels._qc_kernels import _sparse_qc_csr_sub
+
+            block = (32,)
+            grid = (int(math.ceil(X.shape[0] / block[0])),)
+            sparse_qc_csr_sub = _sparse_qc_csr_sub(X.data.dtype)
+            sparse_qc_csr_sub(
+                grid,
+                block,
+                (X.indptr, X.indices, X.data, sums_cells_sub, mask, X.shape[0]),
+            )
+        elif sparse.isspmatrix_csc(X):
+            from ._kernels._qc_kernels import _sparse_qc_csc_sub
+
+            block = (32,)
+            grid = (int(math.ceil(X.shape[1] / block[0])),)
+            sparse_qc_csc_sub = _sparse_qc_csc_sub(X.data.dtype)
+            sparse_qc_csc_sub(
+                grid,
+                block,
+                (X.indptr, X.indices, X.data, sums_cells_sub, mask, X.shape[1]),
+            )
```

Comment:
no need to repeat everything, the only thing that differs between the branches is `X.shape[0]` vs `X.shape[1]` and `_sparse_qc_csr_sub` vs `_sparse_qc_csc_sub`, so the code should look something like the following.

This applies to all code that’s shaped like this.

```py
from ._kernels._qc_kernels import _sparse_qc_csr_sub, _sparse_qc_csc_sub

n, sparse_qc_sub = (
    (X.shape[0], _sparse_qc_csr_sub)
    if sparse.isspmatrix_csr(X) else
    (X.shape[1], _sparse_qc_csc_sub)
)
block = (32,)
sparse_qc_sub(X.data.dtype)(
    grid=(int(math.ceil(n / block[0])),),
    block=block,
    (X.indptr, X.indices, X.data, sums_cells_sub, mask, n),
)
```

Reply:
done